The Sample Space Map (Fundamentals)
Goal: Formalize the process of quantifying randomness.
Scenario: You are analyzing a dataset of customer clicks. A customer can either "Click," "Scroll," or "Exit."
Instructions:
Define the Sample Space $S$ for two consecutive customer actions (e.g., $\{(\text{Click, Click}), (\text{Click, Scroll}), \dots \}$).
Calculate the probability of the event $E$ where the customer clicks at least once.
Write a Python script to simulate rolling two dice $1,000$ times and calculate the experimental probability of the sum being 7.
Why: AI models like Reinforcement Learning rely on understanding every possible state (sample space) to make a move.


## Sample Space for Two Customer Actions

Each action can be:
- Click (C)
- Scroll (S)
- Exit (E)

Total outcomes = 3 × 3 = 9

S = {(C,C), (C,S), (C,E),
     (S,C), (S,S), (S,E),
     (E,C), (E,S), (E,E)}

Event: Customer clicks at least once

Favorable outcomes = 5

Probability = 5/9 = 0.5556


In [3]:
import random

trials = 1000
count_sum_7 = 0

for _ in range(trials):
    dice1 = random.randint(1, 6)
    dice2 = random.randint(1, 6)
    
    if dice1 + dice2 == 7:
        count_sum_7 += 1

experimental_probability = count_sum_7 / trials

print("Number of times sum = 7:", count_sum_7)
print("Experimental Probability:", experimental_probability)


Number of times sum = 7: 180
Experimental Probability: 0.18


## Independent Events

Event A: Flip a coin and get Heads  
$$
P(A) = \frac{1}{2}
$$

Event B: Roll a die and get 6  
$$
P(B) = \frac{1}{6}
$$

Since they are independent:

$$
P(A \cap B) = P(A) \times P(B)
$$

$$
= \frac{1}{2} \times \frac{1}{6}
= \frac{1}{12}
$$


In [8]:
import random

trials = 10000
success = 0

for _ in range(trials):
    coin = random.choice(["H", "T"])
    die = random.randint(1, 6)
    
    if coin == "H" and die == 6:
        success += 1

print("Experimental Probability:", success / trials)

Experimental Probability: 0.0807


## Dependent Events (Without Replacement)

Bag contains:
- 5 Red marbles
- 5 Blue marbles
- Total = 10

Probability first marble is Red:

$$
P(R_1) = \frac{5}{10}
$$

After removing one red:
- Red left = 4
- Total left = 9

$$
P(R_2 | R_1) = \frac{4}{9}
$$

Final Probability:

$$
P(R_1 \cap R_2) = \frac{5}{10} \times \frac{4}{9}
= \frac{2}{9}
$$


In [9]:
import random

trials = 10000
success = 0

for _ in range(trials):
    bag = ["R"]*5 + ["B"]*5
    first = random.choice(bag)
    bag.remove(first)
    
    second = random.choice(bag)
    
    if first == "R" and second == "R":
        success += 1

print("Experimental Probability:", success / trials)


Experimental Probability: 0.2227


### Reflection

In dependent probability, the denominator changed from 10 to 9
because one marble was removed.

This is similar to NLP where the probability of the next word
depends on the previous word.


# Bayes Theorem - Spam Filter Example

Given:

$$
P(Spam) = 0.1
$$

$$
P(Ham) = 0.9
$$

$$
P(Free | Spam) = 0.9
$$

$$
P(Free | Ham) = 0.05
$$

First calculate total probability of "Free":

$$
P(Free) = P(Free|Spam)P(Spam) + P(Free|Ham)P(Ham)
$$


In [10]:
P_spam = 0.1
P_ham = 0.9

P_free_given_spam = 0.9
P_free_given_ham = 0.05


P_free = (P_free_given_spam * P_spam) + (P_free_given_ham * P_ham)


P_spam_given_free = (P_free_given_spam * P_spam) / P_free

print("P(Free) =", P_free)
print("P(Spam | Free) =", P_spam_given_free)


P(Free) = 0.135
P(Spam | Free) = 0.6666666666666667


## Conclusion

The probability that an email is Spam given it contains the word "Free" is:

$$
P(Spam | Free) = 0.667 \approx 66.7\%
$$

This shows how Bayes Theorem updates belief using new evidence.
